### ISSUE TITLE
something wrong with the compilation: Illegal addressing mode (CPU 6502)

### ISSUE THREAD

**Issue by GSoftwareDevelopment** [tebe6502/Mad-Pascal#35]
Suddenly I got this message during compilation:
~~~
rol #$00
"Illegal addressing mode (CPU 6502)"
~~~

This is for the following source line:
~~~pascal
vol_dist[winShift+winPos]:=(vol_dist[winShift+winPos] and $0f) or (v shl 4);
~~~

Compilation result:
~~~asm6502
; optimize OK (instrument.inc), line = 346

	lda WINSHIFT
	add WINPOS
	sta :STACKORIGIN+9
	lda WINSHIFT
	add WINPOS
	tay
	lda adr.VOL_DIST,y
	and #$0F
	sta :STACKORIGIN+10
	lda V
	sta :STACKORIGIN+11
	lda WINPOS+1
	asl :STACKORIGIN+11
	rol @
	sta :STACKORIGIN+STACKWIDTH+11
	rol #$00     <--- it's here
	rol #$00
	asl :STACKORIGIN+11
	rol :STACKORIGIN+STACKWIDTH+11
	asl :STACKORIGIN+11
	rol :STACKORIGIN+STACKWIDTH+11
	asl :STACKORIGIN+11
	rol :STACKORIGIN+STACKWIDTH+11
	lda :STACKORIGIN+10
	ora :STACKORIGIN+11
	ldy :STACKORIGIN+9
	sta adr.VOL_DIST,y
~~~

I'll add that I didn't change anything at this point and this (source) code worked. What is wrong?

**Comment by GSoftwareDevelopment** [comment]
It looks like the compiler can't handle a nested "SHL" instruction.
~~~pascal
modulate[winShift+winPos]:=modulate[winShift+winPos] and $0f or (v shl 4);
~~~

If I move the "SHL" operations outside this area, everything is OK.
~~~pascal
v:=v shl 4;
modulate[winShift+winPos]:=modulate[winShift+winPos] and $0f or v;
~~~

**Comment by tebe6502** [comment]
i can't reproduce this issue
whats type is: vol_dist, winShift, winPos, v ?

**Comment by GSoftwareDevelopment** [comment]
~~~pascal
var
  vol_dist is an array[0..127] of byte;
  winShift,winPos:smallint;
  v:byte;
~~~

I will try a newer version.

**Comment by GSoftwareDevelopment** [comment]
It turns out that I'm using version 1.6.4, and the error is still showing.

**Comment by zbyti** [comment]
> It turns out that I'm using version 1.6.4, and the error is still showing.

` Mad Pascal Compiler version 1.6.5 [2021/05/02] for 6502`

**Comment by zbyti** [comment]
@GSoftwareDevelopment this issue is solved and ready to close with with MP 1.6.5?
